node_modules/oae-principals/tests/test-email.js

```diff
@@ -0,0 +1,280 @@
+                        assert.ok(message.text);
+
+                        // Get the token from the email
+                        var token = message.text.match(/\?token=([a-zA-Z0-9]{32})/)[1];
+                        assert.ok(token);
+
+                        // Verify the token
+                        RestAPI.User.verifyEmailToken(simong.restContext, token, function(err) {
+                            assert.ok(!err);
+
+                            // Assert that the email was verified
+                            RestAPI.User.getMe(simong.restContext, function(err, me) {
+                                assert.ok(!err);
+                                assert.strictEqual(me.emailVerified, true);
+
+                                return callback();
+                            });
+                        });
+                    });
+                });
+            });
+
+            /**
+             * Test that verifies the email tokens are validated
+             */
+            it('verify validation', function(callback) {
+                TestsUtil.generateTestUsers(camAdminRestContext, 1, function(err, users, simong) {
+                    assert.ok(!err);
+
+                    RestAPI.User.verifyEmailToken(simong.restContext, null, function(err) {
+                        assert.strictEqual(err.code, 400);
+
+                        RestAPI.User.verifyEmailToken(simong.restContext, 'non existing token', function(err) {
+                            assert.strictEqual(err.code, 404);
+                            return callback();
+                        });
+                    });
+                });
+            });
+        });
+
+        describe('Resending email tokens', function(callback) {
+
+            /**
+             * Test that verifies parameters are validated
+             */
+            it('verify validation', function(callback) {
+                TestsUtil.generateTestUsers(camAdminRestContext, 2, function(err, users, simong, nico) {
+                    assert.ok(!err);
+                    // Invalid user id
+                    RestAPI.User.resendEmailToken(simong.restContext, 'not a user id', function(err, bla, body) {
+                        assert.strictEqual(err.code, 400);
+
+                        return callback();
+                    });
+                });
+            });
+
+            /**
+             * Test that verifies that unauthorized users cannot resend email tokens
+             */
+            it('verify authorization', function(callback) {
+                TestsUtil.generateTestUsers(camAdminRestContext, 2, function(err, users, simong, nico) {
+                    assert.ok(!err);
+
+                    // Anonymous users cannot resend tokens
+                    RestAPI.User.resendEmailToken(anonymousRestContext, simong.user.id, function(err) {
+                        assert.strictEqual(err.code, 401);
+
+                        // Users cannot resend tokens for other users
```

Comment:
Should we also verify that a tenant admin for tenant X cannot resent a token for a user in tenant Y ?